### PR TITLE
Improved SRAM performance

### PR DESCRIPTION
**Pull request by spikensbror**
This PR improves the way melonDS writes to SRAM.

The current implementation writes directly to the SRAM file on each frame which is very inefficient when it comes to games that write every frame. This PR changes that so that SRAM is flushed to a secondary buffer each "write"-frame and this is then written by a separate thread two seconds after last flush or upon SRAM deinitialization.

This improves the performance of things like saving in the Pokemon games tremendously, especially when running off of non-SSD storage.

As an aside, this fix basically negates the save de-sync when trading Pokemon in some of the Pokemon games (tested in HeartGold and SoulSilver) which means that as long as you can run the emulator above 60 FPS consistently when frame rate is unlocked (fast-forwarding for example), trading will work flawlessly.

I've been running this myself for a while now and have yet to discover any issues.

### CODE REVIEW

File: src/NDSCart_SRAMManager.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        FlushVersion = (FlushVersion + 1 % 255);
+        TimeAtLastFlushRequest = time(NULL);
+        Platform::Mutex_Unlock(SecondaryBufferLock);
+    }
+    
+    void FlushThreadFunc()
+    {
+        for (;;)
+        {
+            usleep(100 * 1000);
```
Comment (RSDuck) [inline src/NDSCart_SRAMManager.cpp:107]: please create a `Sleep` function in Platform.cpp and use QThread::usleep there

**Comment by KostaSaizo7** [comment]
I managed to make 10 trades between Gen4 Pokemon games without an error screen. Well done!

File: src/NDSCart_SRAMManager.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+        Platform::Mutex_Unlock(SecondaryBufferLock);
+    }
+
+    void RequestFlush()
+    {
+        Platform::Mutex_Lock(SecondaryBufferLock);
+        printf("NDS SRAM: Flush requested\n");
+        memcpy(SecondaryBuffer, Buffer, Length);
+        FlushVersion = (FlushVersion + 1 % 255);
```
Comment (RSDuck) [inline src/NDSCart_SRAMManager.cpp:98]: modulo has precedence over addition

File: src/NDSCart_SRAMManager.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        Platform::Mutex_Lock(SecondaryBufferLock);
+
+        strncpy(Path, path, 1023);
+        Path[1023] = '\0';
+        
+        Buffer = buffer;
+        Length = length;
+        
+        SecondaryBuffer = new u8[length];
```
Comment (RSDuck) [inline src/NDSCart_SRAMManager.cpp:83]: this can leak memory

**Review by RSDuck — CHANGES_REQUESTED** [review]
great addition!

File: src/NDSCart_SRAMManager.cpp (hunk 2 above)
Comment (spikensbror) [reply src/NDSCart_SRAMManager.cpp:98]: I think this is due to a "typo", the parenthesis should only encapsulate the `FlushVersion + 1` part. I'll remove it though, I think I did this before the final version when messing around with potential solutions.

File: src/NDSCart_SRAMManager.cpp (hunk 3 above)
Comment (spikensbror) [reply src/NDSCart_SRAMManager.cpp:90]: Should be OK now!

File: src/NDSCart_SRAMManager.cpp (hunk 2 above)
Comment (spikensbror) [reply src/NDSCart_SRAMManager.cpp]: Should be fixed.

File: src/NDSCart_SRAMManager.cpp (hunk 1 above)
Comment (spikensbror) [reply src/NDSCart_SRAMManager.cpp]: Should be fixed.

File: src/NDSCart_SRAMManager.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    void FlushThreadFunc()
+    {
+        for (;;)
+        {
+            Platform::Sleep(100 * 1000); // 100ms
+
+            if (!FlushThreadRunning) return;
+            
+            // We debounce for two seconds after last flush request to ensure that writing has finished.
+            if (TimeAtLastFlushRequest == 0 || time(NULL) - TimeAtLastFlushRequest < 2)
```
Comment (RSDuck) [inline src/NDSCart_SRAMManager.cpp:119]: this usage of time is unportable. Though idk how much this matters.

Comment (spikensbror) [reply src/NDSCart_SRAMManager.cpp:119]: I used `time(NULL)` as I saw it being used elsewhere as well (RTC.cpp, frontend/qt_sdl/main.cpp). Let me know what you want here and I'll accommodate!

Comment (spikensbror) [reply src/NDSCart_SRAMManager.cpp]: Nevermind, I see what you're saying. Should be fixed in the latest push!

**Comment by Gamer654** [comment]
Sorry but how do I apply this patch to the emulator

**Comment by fancythedeveloper** [comment]
Just do git pull and compile the latest version.

**Comment by WaluigiWare64** [comment]
Or you could download the latest builds from GitHub Actions.

**Comment by Gamer654** [comment]
> 
> 
> Just do git pull and compile the latest version.

how? I'm new to github and don't know how too.

**Comment by WaluigiWare64** [comment]
> 
> 
> > Just do git pull and compile the latest version.
> 
> how? I'm new to github and don't know how too. I mean, I did compile it, but don't know where the directory is stored or where to find it on my C drive.

@Gamer654 look at my above comment.

**Comment by Gamer654** [comment]
> 
> 
> > > Just do git pull and compile the latest version.
> > 
> > 
> > how? I'm new to github and don't know how too. I mean, I did compile it, but don't know where the directory is stored or where to find it on my C drive.
> 
> @Gamer654 look at my above comment.

I saw that, but I thought it was the same thing like compiling it. I don't know what Github actions are, so how do I do it or where do I download it?

**Comment by WaluigiWare64** [comment]
Go to the README, click on the badge that says your OS on, then click the topmost entry in the list. Scroll down to the artifacts section, and there you can find the latest melonDS from the CI.

**Comment by Gamer654** [comment]
> 
> 
> Go to the README, click on the badge that says your OS on, then click the topmost entry in the list. Scroll down to the artifacts section, and there you can find the latest melonDS from the CI.

pretty sure I got the right one, however it gives an error.

**Comment by WaluigiWare64** [comment]
@Gamer654 download the latest artifact, I just fixed that issue.

**Pull request merged**